dev_tools/oneapi/convert_source.sh

```diff
@@ -0,0 +1,429 @@
+replace_regex="${replace_regex};s/thread::/thread_t::/g"
+replace_regex="${replace_regex};s/bitonic_sort/bitonic_sort_t/g"
+replace_regex="${replace_regex};s/reduction_array/reduction_array_t/g"
+replace_regex="${replace_regex};s/cuda\/components\/thread_ids.cuh/trick\/thread_ids.hpp/g"
+replace_regex="${replace_regex};s/cuda\/components\/cooperative_groups.cuh/trick\/cooperative_groups.hpp/g"
+replace_regex="${replace_regex};s/cuda\/components\/sorting.cuh/trick\/sorting.hpp/g"
+replace_regex="${replace_regex};s/cuda\/components\/reduction.cuh/trick\/reduction.hpp/g"
+replace_regex="${replace_regex};s/CUH_/DP_HPP_/g"
+replace_regex="${replace_regex};s/(.*template.*)(Config * [a-zA-Z0-9_]*) * = *[a-zA-Z0-9_:]*(\(+[^\)]*\))?/\1\2/g"
+# keep using original xxx.sync(); ->xxx;//.sync();
+replace_regex="${replace_regex};s/(\.sync\(\);)/;\/\/.sync()/g"
+# template macro(); lead std::length_error
+replace_regex="${replace_regex};s/(template GKO.*;)/\/\/ \1/g"
+sed -i -E "${replace_regex}" "${OUTPUT_FILE}"
+if grep -Eq "dim3" ${OUTPUT_FILE}; then
+    # Found
+    sed -i '1 i#include "trick/dim3_t.hpp"' ${OUTPUT_FILE}
+fi
+# add the cooperative group header according to group:: because some sources forget to add it
+if grep -Eq "group::" ${OUTPUT_FILE}; then
+    # Found
+    sed -i '1 i#include "trick/cooperative_groups.hpp"' ${OUTPUT_FILE}
+fi
+# exit 0
+
+OUTPUT_FOLDER="output"
+if [[ "${VERBOSE}" == 1]]; then
+    echo "The dpct calling:"
+    echo "dpct --extra-arg=\"-std=c++14\" --extra-arg=\"-I ${ROOT_DIR}\" --extra-arg=\"-I ${ROOT_DIR}/include\" --extra-arg=\"-I ${ROOT_DIR}/dev_tools/oneapi\" --extra-arg=\"-I ${GTEST_HEADER_DIR}\" --cuda-include-path=\"${CUDA_HEADER_DIR}\" --format-range=none ${OUTPUT_FILE} --out-root=${OUTPUT_FOLDER}"
+fi
+
+# Delete output/source.cpp
+rm "${OUTPUT_FOLDER}/${OUTPUT_FILE}"
+rm "${OUTPUT_FOLDER}/${OUTPUT_FILE}.dp.cpp"
+
+dpct --extra-arg="-std=c++14" --extra-arg="-I ${ROOT_DIR}" --extra-arg="-I ${ROOT_DIR}/include" --extra-arg="-I ${ROOT_DIR}/dev_tools/oneapi" --extra-arg="-I ${GTEST_HEADER_DIR}" --cuda-include-path="${CUDA_HEADER_DIR}" --format-range=none ${OUTPUT_FILE} --out-root=${OUTPUT_FOLDER}
+
+dpct_file=""
+if [ -f "${OUTPUT_FOLDER}/${OUTPUT_FILE}.dp.cpp" ]; then
+    dpct_file="${OUTPUT_FOLDER}/${OUTPUT_FILE}.dp.cpp"
+elif [ -f "${OUTPUT_FOLDER}/${OUTPUT_FILE}" ]; then
+    dpct_file="${OUTPUT_FOLDER}/${OUTPUT_FILE}"
+else
+    echo "No file"
+    exit 1
+fi
+
+cp "${dpct_file}" "${dpct_file}_bkp"
+if [[ "${VERBOSE}" == 1 ]]; then
+    echo "the dpct result ${dpct_file}_bkp"
+    echo "recover the temporary change in the file ${dpct_file}"
+fi
+# global reverse fix
+# dim3_t -> dim3
+# this_thread_block_t -> this_thread_block
+# tiled_partition_t -> tiled_partition
+# thread_id
+# cooperative_group
+# replace_regex="1 i\#include <CL/sycl.hpp>"
+# replace_regex="s/cuda/dpcpp/g"
+# replace_regex="${replace_regex};s/#include \\\"common/#include \\\"dpcpp_code/g"
```

Comment:
I think the previous issues should be documented somewhere,
in a markdown file or something. Maybe detailed a little if you have an
example of failures?